Expand Support To Unbound

evanwporter opened the pull request:
This will expand support to Pokemon Unbound. This will be made possible with #80

evanwporter commented:
We should make an abstract class for use by RR and UB. This will promote code reuse.

evanwporter commented:
@andrewbenington 

Unbound is looking really good. Several issues however:

* Game of Origin
* Sprites. Unbound has more forms than RR (like all the forms of vivillion). Will need to selectively copy over the forms from [here](https://github.com/Skeli789/Dynamic-Pokemon-Expansion/tree/master/graphics/frontspr).
  * A way do this is to compare the current sprite list and the target sprite list (link above) and remove all duplicate files (names are different though so will need to read file data)
* Unbound logo. I added it as a webp file and a png file. But I can't remember how its logos are accessed for plugins. Also will need to add bg to the png.

evanwporter commented:
![image](https://github.com/user-attachments/assets/480b95bb-e228-4281-b67b-13b92940ff75)

Looks clean...

evanwporter commented:
[80% unbound.zip](https://github.com/user-attachments/files/18194780/80.unbound.zip)

By the way been doing this save for tests which I found here => https://projectpokemon.org/home/files/file/4911-pokemon-unbound-save-file/

evanwporter commented:
@andrewbenington 

Sprites are done

Now we just need to transfer restrictions & Items

evanwporter reviewed (commented):
@andrewbenington

I finished the Unbound stuff. There may be a couple issues here and there with regards to the the correct sprites and the species name, but I they won't break anything and a tweak or two to the `conversions/` will resolve it in a jiffy.

Let me know what you think. 

When this is merged you should post OpenHome to `r/unbound` and `r/radicalred`. Mention how OpenHome can be used as an alternative to Unbound Cloud, except it has support for `Radical Red` & `Unbound` & every other mainline pokemon game (Unbound Cloud only supports Unbound).

andrewbenington commented:
@evanwporter Looks good, just a couple image-related things. I don't want to add any new sprites to the codebase at this point, because it *really* takes up a lot of space and clogs up the git history. Even just moving the radical red sprites like you've done here has doubled the space those sprites take up in the .git files, because git is treating them like new files with their own history.

When I audited the git history to remove the old sprites it was a huge pain, but it really brought down the time/space needed to clone the repository. I don't want to have to do that again.

Right now my priority is getting a full release of the tauri version out, which should be any day now, but after that I will work on making it possible to download asset packs like Unbound sprites separately without them taking up space here. I would save the unbound sprites somewhere else on your computer to make it quick and easy to make them an asset pack when I get around to that.

I would love if more Radical Red/Unbound fans used OpenHome! But it should probably be you to post about it on those subreddits because I've never played them and I don't want to come off as a self-promoting poser 😂

andrewbenington commented on src/pokemon/FileTypeSelect.tsx:
```diff
@@ -55,6 +55,8 @@ const FileTypeSelect = (props: FileTypeSelectProps) => {
   const [, , getEnabledSaveTypes] = useContext(AppInfoContext)
 
   const supportedFormats = useMemo(() => {
+    console.log(getEnabledSaveTypes())
```
remove console log

evanwporter commented:
@andrewbenington 

>Looks good, just a couple image-related things. I don't want to add any new sprites to the codebase at this point, because it really takes up a lot of space and clogs up the git history. Even just moving the radical red sprites like you've done here has doubled the space those sprites take up in the .git files, because git is treating them like new files with their own history.

A few images (less than 10) have been added but they take up a tiny bit. The 1000+ images you see in changed files are from when I changed the name `spites/rr` => `sprites/cfru`. Changed this back and the files shrunk from 1000+ files changed to 34 (at the time of writing). Also I've been making crazy changes left and right to the sprites. Which absolutely clog up the git history but this won't be an issue, as long as you click `squash and merge` and not just merge when this branch is merged. Unbound and rr pull from the same sprite folder.

andrewbenington commented:
> A few images (less than 10) have been added but they take up a tiny bit. The 1000+ images you see in changed files are from when I changed the name `spites/rr` => `sprites/cfru`. Changed this back and the files shrunk from 1000+ files changed to 34 (at the time of writing). Also I've been making crazy changes left and right to the sprites. Which absolutely clog up the git history but this won't be an issue, as long as you click `squash and merge` and not just merge when this branch is merged. Unbound and rr pull from the same sprite folder.

@evanwporter Ah ok that's fine then. Btw in the future if you want to move or rename a file/directory, you can use `git mv <old name/location> <new name/location>` and git will consider it a rename instead of a deleted file and new file. That was helpful when I moved everything out of the old `renderer` folder. But yeah as long as it's squashed before the merge we don't have to worry about those commits on this branch

andrewbenington commented on public/sprites/rr/.png:
This file should probably be deleted

evanwporter commented:
@andrewbenington 

Last thing to do before we merge is the met location. The met location is registering as RadicalRed. Imo this is unavoidable since both roms have the same met-game metadata, so its either one or the other. Maybe someone smarter than me in the future will find a solution? If your okay with both roms occupying the same met game and met data then I think its ready to merge. Other issues can be resolved on a rolling basis like we did for #91.

andrewbenington commented on src/types/SAVTypes/unbound/G3UBSAV.ts:
```diff
@@ -0,0 +1,70 @@
+
+  static saveTypeAbbreviation = 'Unbound'
+  static saveTypeName = 'Pokémon Unbound'
+  static saveTypeID = 'G3RRSAV'
```
```suggestion
  static saveTypeID = 'G3UBSAV'
```

evanwporter replied:
Really good catch here!

andrewbenington commented on src/types/SAVTypes/unbound/G3UBSAV.ts:
```diff
@@ -0,0 +1,70 @@
+import { PluginSAV } from '../SAV'
+import PK3UB from './PK3UB'
+
+const UB_TRANSFER_RESTRICTIONS: TransferRestrictions = {
+  maxDexNum: NationalDex.Enamorus,
+  // excludedForms: {
+  // ...CapPikachus,
+  // [NationalDex.Pichu]: [SPIKY_EAR],
+  // [NationalDex.Dialga]: [ORIGIN],
+  // [NationalDex.Palkia]: [ORIGIN],
+  // },
+}
```
```suggestion
const UB_TRANSFER_RESTRICTIONS: TransferRestrictions = {
  maxDexNum: NationalDex.Enamorus,
}
```

andrewbenington commented on src/types/SAVTypes/unbound/G3UBSAV.ts:
```diff
@@ -0,0 +1,70 @@
+
+    return gameCode === 255
+
+    // const securityKey = bytesToUint32LittleEndian(firstSectionBytes, FRLG_SECURITY_OFFSET)
+    // const securityKeyCopy = bytesToUint32LittleEndian(firstSectionBytes, FRLG_SECURITY_COPY_OFFSET)
+
+    // console.log('UNBOUND SEC KEY', securityKey, securityKey === 0, securityKey !== securityKeyCopy)
+
+    // // Radical Red seems to have the security key set to 0, which has a 1 in 4.2 billion
+    // // chance to happen in vanilla FireRed (if it can even be 0 at all)
+    // return securityKey === 0 || securityKey !== securityKeyCopy
+  }
```
```suggestion
```

andrewbenington commented on src/types/SAVTypes/cfru/PK3CFRU.ts:
```diff
@@ -0,0 +1,527 @@
+        this.metLocationIndex = other.metLocationIndex ?? FIRERED_IN_GAME_TRADE
+        this.internalMetLocationIndex = FIRERED_IN_GAME_TRADE
+      }
+
```
I would move this code out of the super class and into the subclasses, after a `super()` call. This is why you're having issues tracking the `unbound` origin

andrewbenington commented on src/types/SAVTypes/cfru/PK3CFRU.ts:
```diff
@@ -0,0 +1,527 @@
+
+      // OHPKM will not copy this field if the mon was already being tracked before being
+      // transferred to Radical Red
+      this.pluginOrigin = this.gameOfOrigin === INTERNAL_ORIGIN_FROM_RR ? 'radical_red' : undefined
```
same here

andrewbenington reviewed (dismissed):
There are also a bunch of other commented out sections of code I assume you can remove.

If you're having issues with the unbound origin even after moving that pluginOrigin code into the subclasses I can take a look. There shouldn't be any issues with that as long as the save file itself has been determined to be Unbound.

evanwporter commented:
>If you're having issues with the unbound origin even after moving that pluginOrigin code into the subclasses I can take a look. There shouldn't be any issues with that as long as the save file itself has been determined to be Unbound.

Well what if a mon originates in RR but is moved to Unbound, wouldn't it then say its from RR

andrewbenington commented:
> > If you're having issues with the unbound origin even after moving that pluginOrigin code into the subclasses I can take a look. There shouldn't be any issues with that as long as the save file itself has been determined to be Unbound.
> 
> Well what if a mon originates in RR but is moved to Unbound, wouldn't it then say its from Unbound

Nope, that mon's OHPKM file will already have its pluginOrigin set to `radical_red` from when it was moved out of Radical Red. The `pluginOrigin` field is only set once, and never updated

evanwporter commented:
Oh okay. Yeah how might that work then?

evanwporter commented:
@andrewbenington 

Nvm I got it to work. Looks clean.

![image](https://github.com/user-attachments/assets/0496330b-b81a-494a-b650-26be51cfe2b8)

andrewbenington commented:
> @andrewbenington
> 
> Nvm I got it to work. Looks clean.
> 
> ![image](https://private-user-images.githubusercontent.com/115374841/397961970-0496330b-b81a-494a-b650-26be51cfe2b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ5MjUyNzIsIm5iZiI6MTczNDkyNDk3MiwicGF0aCI6Ii8xMTUzNzQ4NDEvMzk3OTYxOTcwLTA0OTYzMzBiLWI4MWEtNDk0YS1iNjUwLTI2YmU1MWNmZTJiOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyM1QwMzM2MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZDA4NDY5NzdiNmNlMTZiNjNjZmE0ZDk1YWIyODk2YWI4NjRmYTIwZmEwYWUwMWI0MWM3NTNiY2EwYmQzY2MyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.OA6Y3qXZd3nti2HmNx-7Y4B_c-Oj8ZLDG1QSpdZPyqA)

Looks great. I like the dropdown color you picked too.

I'm going to do some testing on my own in the coming days and then I'll merge this. Apologies in advance if this takes a bit; I'm pretty busy this week with the holidays
